pkg/api/v1/resources/resource_interface.go

```diff
@@ -44,8 +44,16 @@ func ProtoCast(res Resource) (ProtoResource, error) {
 
 type InputResource interface {
 	Resource
+	// Deprecated: statuses are now written to a ReporterStatus, use GetNamespacedStatus() or
+	// GetReporterStatus() instead.
 	GetStatus() *core.Status
 	SetStatus(status *core.Status)
+	HasStatus() bool
+	GetReporterStatus() *core.ReporterStatus
+	SetReporterStatus(status *core.ReporterStatus)
+	HasReporterStatus() bool
+	GetNamespacedStatus() *core.Status
```

Comment:
What's the purpose of expanding the interface for HasReporterStatus and UpsertReporterStatus? I'm a little confused by deprecating methods on the existing interface to prefer new ones with the same signature

Reply:
`HasReporterStatus` is a helper for determining if the `status_oneof` contains a `ReporterStatus`, as opposed to a `Status`.  It's necessity is debatable, and usages could probably be replaced by `.GetReporterStatus() != nil`, but I kind of like the Has...() syntax as a subtle reminder that it's a oneof

`UpsertReporterStatus` is a helper for obtaining the value in the map (a `Status`) that corresponds to the `POD_NAMESPACE` of the current environment.  This one saves us from having repeated logic scattered across the codebase which calls `GetReporterStatus()`, then `os.Getenv("POD_NAMESPACE")` to determine the proper key to use to add/update a `Status` value in the `ReporterStatus.Statuses` map.

`GetStatus` is deprecated, because going forward it will be incorrect to use that function 99.9% of the time. The only time it would be valid is when we're reading a resource that has a `Status` that has not yet been written as a `ReporterStatus` (i.e. after upgrade).

Reply:
Sorry I think my comment was a little misleading. I meant the GetNamespacedStatus and UpsertReporterStatus methods which deprecate the GetStatus, SetStatus methods, but share the same signature. I like the `Has` methods as well! 

I had imagined the interface for a resource would be Get/Set for both Status and ReporterStatus, but that the SetStatus implementation would change to write to the ReporterStatus first, and if not possible, fallback to the Status field, and likewise, the GetStatus implementation would change to read from the ReporterStatus field first, and fallback on the status field. 

Perhaps this isn't possible / doesn't make sense, though.

Reply:
We need `GetStatus()` to return nil when the `status_oneof` contains a `NamespacedStatuses` (formerly ReporterStatus) so that we can detect what's in the oneof. HasStatus() and HasNamespacedStatuses() have been removed since the last comment